cms/search/publishers.py

```diff
@@ -0,0 +1,163 @@
+    def _publish(self, channel: str | None, message: dict) -> None:
+        """Each child class defines how to actually send/publish
+        the message (e.g., Kafka, logging, etc.).
+        """
+
+    @abstractmethod
+    def get_created_or_updated_channel(self) -> str | None:
+        """Provide the channel (or other necessary routing key) for created/updated.
+        This can be a no-op or empty string for some implementations.
+        """
+
+    @abstractmethod
+    def get_deleted_channel(self) -> str | None:
+        """Provide the channel (or other necessary routing key) for deleted messages.
+        This can be a no-op or empty string for some implementations.
+        """
+
+    def _construct_message_for_create_update(self, page: "Page") -> dict:
+        """Build a dict that matches the agreed metadata schema for 'created/updated'.
+
+        resource_metadata.yml:
+           - StandardPayload: requires {uri, title, content_type}, optional fields
+           - ReleasePayload: extends StandardPayload with release-specific fields
+        """
+        # Common fields for StandardPayload (also part of ReleasePayload)
+        # The schema requires at minimum: uri, title, content_type
+        message = {
+            "uri": page.url_path,
+            "content_type": page.search_index_content_type,
+            "release_date": (
+                page.release_date.isoformat()
+                if page.search_index_content_type == "release" and getattr(page, "release_date", None)
+                else None
+            ),
+            "summary": get_text_for_indexing(force_str(page.summary)),
+            "title": page.title,
+            "topics": getattr(page, "topic_ids", []),
+        }
+
+        # If it's a Release, we add the extra fields from ReleasePayload
+        if page.search_index_content_type == "release":
+            message.update(self._construct_release_specific_fields(page))
+        return message
+
+    def _construct_release_specific_fields(self, page: "Page") -> dict:
+        """Constructs and returns a dictionary with release-specific fields."""
+        release_fields = {
+            "finalised": page.status in ["CONFIRMED", "PROVISIONAL"],
+            "cancelled": page.status == "CANCELLED",
+            "published": page.status == "PUBLISHED",
+            "date_changes": [],
+        }
+
+        # If page.changes_to_release_date exists, add the first item's reason & previous date
+        if hasattr(page, "changes_to_release_date") and page.changes_to_release_date:
+            first_change = page.changes_to_release_date[0].value
+            release_fields["date_changes"].append(
+                {
+                    "change_notice": first_change.get("reason_for_change"),
+                    "previous_date": str(first_change.get("previous_date").isoformat())
+                    if first_change.get("previous_date")
+                    else None,
+                }
+            )
```

Comment:
This is only adding the first change. There might be multiple changes to the release date. The previous date is mandatory so you don't need the `if` statement.
Need to update tests.
```suggestion
            release_fields["date_changes"] = [
                {
                    "change_notice": change.value.get("reason_for_change"),
                    "previous_date": change.value.get("previous_date").isoformat(),
                }
                for change in page.changes_to_release_date
            ]
```

Reply:
Done!